add CLI option to syntax check script

bahamas10 opened the pull request:
original PR: https://github.com/joyent/node/pull/9447
original discussion: https://github.com/joyent/node/issues/9426

examples

```
$ cat ./good.js
var foo = 'bar';
$ ./out/Release/iojs -c ./good.js
$ echo $?
0
```

```
$ cat ./bad.js
var foo bar;
$ ./out/Release/iojs -c ./bad.js
var foo bar;
        ^^^

SyntaxError: Unexpected identifier
    at startup (node.js:99:13)
    at node.js:961:3
$ echo $?
1
```

tests pass

```
$ ./out/Release/iojs test/parallel/test-cli-syntax.js 
calling /Users/dave.eddy/dev/node-1/out/Release/iojs -c /Users/dave.eddy/dev/node-1/test/fixtures/syntax/good_syntax.js
ok
calling /Users/dave.eddy/dev/node-1/out/Release/iojs --check /Users/dave.eddy/dev/node-1/test/fixtures/syntax/good_syntax.js
ok
calling /Users/dave.eddy/dev/node-1/out/Release/iojs -c /Users/dave.eddy/dev/node-1/test/fixtures/syntax/good_syntax_shebang.js
ok
calling /Users/dave.eddy/dev/node-1/out/Release/iojs --check /Users/dave.eddy/dev/node-1/test/fixtures/syntax/good_syntax_shebang.js
ok
calling /Users/dave.eddy/dev/node-1/out/Release/iojs -c /Users/dave.eddy/dev/node-1/test/fixtures/syntax/bad_syntax.js
ok
calling /Users/dave.eddy/dev/node-1/out/Release/iojs --check /Users/dave.eddy/dev/node-1/test/fixtures/syntax/bad_syntax.js
ok
calling /Users/dave.eddy/dev/node-1/out/Release/iojs -c /Users/dave.eddy/dev/node-1/test/fixtures/syntax/bad_syntax_shebang.js
ok
calling /Users/dave.eddy/dev/node-1/out/Release/iojs --check /Users/dave.eddy/dev/node-1/test/fixtures/syntax/bad_syntax_shebang.js
ok
```

bnoordhuis commented on src/node.js:
```diff
@@ -91,6 +91,15 @@
+        if (process._syntax_check_only != null) {
+            var vm = NativeModule.require('vm');
+            var fs = NativeModule.require('fs');
+            var source = fs.readFileSync(process.argv[1], 'utf-8').replace(/^#.*/, '');
```
I see a couple of issues here:

1. It won't work for filenames without an extension (`node app` vs. `node app.js`)
2. It won't skip the UTF-8 BOM (byte order marker) like `require()` does.
3. The regular expression doesn't match the one in lib/module.js that matches the string `#!`.

It's probably unavoidable that fixing 2 and 3 results in some duplicated code but it should come with regression tests that ensure the implementations won't diverge.

bnoordhuis commented on test/parallel/test-cli-syntax.js:
```diff
@@ -0,0 +1,55 @@
+var assert = require('assert');
```
Can you add `'use strict'` here and use `const` for the imports?  Didn't `make test` complain about that?

bnoordhuis commented on test/parallel/test-cli-syntax.js:
```diff
@@ -0,0 +1,55 @@
+    assert.equal(c.stdout, '', 'stdout produced');
+    assert(c.stderr, 'stderr not produced');
+    assert.equal(c.status, 1, 'code == ' + c.status);
+    console.log('ok');
```
Can you remove the console.log statements in this file?  Rule of thumb is that tests should be (mostly) quiet on success.

bnoordhuis commented:
Left some comments.  Basic premise of this PR seems alright to me.

bahamas10 commented:
Thanks @bnoordhuis for the review.

1. I've updated the code to use `Module._resolveFilename` on `process.argv[1]` to fix the `node app` vs `node app.js` problem.
2. I've exposed `stripBOM` as `Module._stripBOM` which is used by this code in `node.js` (to avoid repeated code
3. updated regex to match that in `lib/module.js`
4. updated tests to use `const` and `'use strict';`... it turns out I was testing `joyent/node` and not `nodejs/node` accidentally before... so i've made sure this all passes `make test` in this repo
5. removed all logging on success in the test supplied

Additionally I've added to test for `file` vs `file.js`, as well as files that don't exist.

thefourtheye commented on src/node.cc:
```diff
@@ -2843,6 +2844,11 @@ void SetupProcessObject(Environment* env,
 
+  // -c, --check
+  if (syntax_check_only) {
+    READONLY_PROPERTY(process, "_syntax_check_only", True(env->isolate()));
```
I would prefer using consistent camel casing. I am not sure if we can fix the `_print_eval` above. cc @bnoordhuis

bahamas10 replied:
I'd be happy to change this if everyone agrees this should be done - I just tried to match the style I saw in this file.

bnoordhuis replied:
It's consistent with `_print_eval`.  I'm fine with it.

thefourtheye replied:
Oh okay then. I was just worried that we would introduce non camel case identifiers in to the js land.

trevnorris replied:
Is this needed? the syntax check won't actually launch the script so there will never be access to `process._syntax_check_only`.

trevnorris replied:
wait. nm. was thinking from the users' perspective. forgot about internal tooling.

thefourtheye commented on test/parallel/test-cli-syntax.js:
```diff
@@ -0,0 +1,76 @@
+
+  // loop each possible option, `-c` or `--check`
+  syntax_args.forEach(function(args) {
+    var _args = args.concat([file]);
```
It can simply be written as `args.concat(file);`

bahamas10 replied:
no it can't - concat does not modify the array in-place.  see:

``` js
> var a = ['foo']
> var b = ['bar']
> a.concat(b)
[ 'foo', 'bar' ]
> a
[ 'foo' ]
> b
[ 'bar' ]
```

If it *did* modify it in-place, this loop wouldn't work because every iteration the array would grow.

bahamas10 replied:
You know, upon re-reading this I see that you probably weren't talking about removing the `var _args =` portion of this, but were just referring to `file` vs `[file]` :)

Didn't know that could be done - changing it now, thanks!

thefourtheye commented on test/parallel/test-cli-syntax.js:
```diff
@@ -0,0 +1,76 @@
+
+    // no output should be produced
+    assert.equal(c.stdout, '', 'stdout produced');
+    assert(c.stderr, 'stderr not produced');
```
I feel that checking the exact error messages will be good, here.

bahamas10 replied:
The error messages look like this

```
$ ./out/Release/iojs -c ./bad.js
var foo bar;
        ^^^

SyntaxError: Unexpected identifier
    at startup (node.js:99:13)
    at node.js:961:3
```

Checking them would be difficult, because they have line numbers that are subject to change in the future.  We *could* use regex to test them, but that seems unstable and like it could break in the future.

thefourtheye replied:
Correct. We should at least check if the srderr has `SyntaxError: Unexpected identifier`. Otherwise this assert is meaningless.

rvagg replied:
@bahamas10 can you add an assert in here to expand the test a bit please?

thefourtheye commented on lib/module.js:
```diff
@@ -441,6 +441,8 @@ function stripBOM(content) {
   return content;
 }
 
+Module._stripBOM = stripBOM;
```
Should this really be exposed like this? Is there a better option?

bnoordhuis replied:
I suggest moving stripBOM to lib/internal/module.js (to be created) and then require() it in lib/module.js and src/node.js.

bahamas10 replied:
@bnoordhuis agreed.  I exposed it as-is to avoid duplicating code - I'll create the new module and push a new commit up in a bit.

alexrepin commented:
echo "$"

bnoordhuis commented on test/parallel/test-cli-syntax.js:
```diff
@@ -0,0 +1,76 @@
+var node = process.argv[0];
+
+// test both sets of arguments that check syntax
+var syntax_args = [
```
Can you spell this as `syntaxArgs`?  snake_case is not entirely prohibited in the test suite but the convention is camelCase.

bahamas10 replied:
ack. will do in next revision

rvagg commented:
Hey @bahamas10, welcome!

This does seem like something that could easily be achieved by a userland module, surely there's already something in npm that does this, so what's the case for putting it in core?

bahamas10 commented:
Hey @rvagg 

The indirect answer is here https://github.com/joyent/node/issues/9426

But to be more specific, to syntax check scripts:

- `bash -n somescript`
- `ruby -c somescript.rb`
- `perl -c `somescript.pl`

etc.

Most interpreted language interpreters support some way of saying "are the contents of the file given valid for interpretation?".  This does not check style nor does it lint the content - just syntax.

However, I realize the argument that "a bunch of people are doing it" is not sufficient.

The node core is *already* the source of truth when it comes to compiling JavaScript source and determining if it is valid.  For example, an `npm` module would be nothing more than a thin wrapper around the `vm` module and `new vm.Script(...)`.  Since the node core is already doing the heavy lifting, and it itself contains the logic to determine the validity of JavaScript source code, it alone should be responsible for it.

As an example, imagine the following script (downloaded from `npm` or wherever) called `syntax-check`

``` js
#!/usr/bin/env node
var vm = require('vm');
var fs = require('fs');
var source = fs.readFileSync(process.argv[2], 'utf8').replace(/^#.*/, '');
new vm.Script(source, {displayErrors: true});
```

You could then call this script in the following way to achieve *almost* the same results as this PR

```
$ ./syntax-check syntax-check
$ echo $?
0
```

However, this only checks the validity of it against whichever `node` binary is available first in the users `$PATH` (because of the shebang).  So test other versions of node a user must run something like

```
$ ~/dev/node/out/Release/node "$(npm config get prefix)/bin/syntax-check" script.js
```

This requires a lot of knowledge about the underlying system, and also forces a specific interpreter for the script, going against what is in the shebang (ie. what the user intended it to be executed with) - thus running a script downloaded from `npm` for a specific version of node with a different version of node.

I think it makes the most sense to cut out anything in the middle - any 3rd party script - and just ask the `node` binary itself "can you run this?"

bahamas10 commented:
1. moved `stripBOM` to `internal/module.js` and use it in both `lib/module.js` and `src/node.js`
2. `syntax_args` => `syntaxArgs`
3. `args.concat([file])` => `args.concat(file)`

`make test` passes

bnoordhuis commented:
The commit log should conform to the guidelines from CONTRIBUTING.md but otherwise LGTM.  Let's see what the CI thinks: https://jenkins-iojs.nodesource.com/job/node-test-pull-request/104/

bahamas10 commented:
updated commit message based on CONTRIBUTING.md guidelines.

rvagg commented:
OK, I'm -0 on this, I won't hold up merging if other collaborators feel strongly enough that this is a good thing to have

bahamas10 commented:
(merge conflicts resolved and rebased with latest master)

bahamas10 commented:
@rvagg I don't feel comfortable about this being merged if you are having reservations about it.  Is there anything specific about this you don't like? or is it an overall negative feeling about adding something extra to the node core?

If you're available I'm happy to talk about this and make sure we are both on the same side with regards to this change.

bnoordhuis commented:
I didn't comment on it last time but FWIW, it seems like a reasonable addition to me with precedent in other scripting environments.  Putting it in a npm package makes it markedly less convenient.

rvagg commented:
yeah, sorry for not responding earlier, this is fine by me, the fact that other runtimes have the same functionality makes this easier to swallow

rvagg commented:
https://ci.nodejs.org/job/node-test-pull-request/254/ again, the last one looked pretty flaky

thefourtheye commented on src/node.js:
```diff
@@ -92,6 +92,22 @@
+        // check if user passed `-c` or `--check` arguments to Node.
+        if (process._syntax_check_only != null) {
+          var vm = NativeModule.require('vm');
+          var fs = NativeModule.require('fs');
```
Use `const` wherever the values are assigned once and never changed again.

rvagg replied:
this could go either way cause it's not at all consistent in the file, we might be shifting that way but there's only one use of `const` amongst the 20 `NativeModule.require` calls already in this file, someone's probably going to do a round of cleanup at some point and this change could be done then

thefourtheye replied:
@rvagg Oh, okay. I thought if we allowed the new code being landed to follow this rule strictly, then we just have to cleanup only the old code.

rvagg commented:
all green, https://ci.nodejs.org/job/node-test-commit/506/, that's pretty good given the recent state of our suite, lgtm

evanlucas commented:
LGTM

rvagg commented:
@nodejs/collaborators any objections to adding this non-trivial functionality? I'm at +1 on this and will merge unless someone sticks their head up to object.

indutny commented:
I don't mind, sounds reasonable

domenic commented:
When I saw the subject line in my email I was against it but after reading https://github.com/nodejs/node/pull/2411#issuecomment-132263887 I am for it.

mscdex commented:
+1

targos commented:
+1

i5ting commented:
+1

ChALkeR commented:
+0

thefourtheye commented on test/parallel/test-cli-syntax.js:
```diff
@@ -0,0 +1,76 @@
+
+const common = require('../common');
+
+var node = process.argv[0];
```
Use [`process.execPath`](https://nodejs.org/api/process.html#process_process_execpath)

rvagg replied:
@bahamas10 could you change this please?

Fishrock123 commented:
Seems fine.

yosuke-furukawa commented:
+1

rvagg commented:
@bahamas10 I see 2 outstanding items, pointed them out inline, then I'll merge this and we should be able to get it into v4.x

bahamas10 commented:
thanks @thefourtheye for the problems you've spotted.  @rvagg i've force pushed a new commit with the changes... here is the intermediate diff to easily spot the changes

``` diff
$ gd
diff --git a/test/parallel/test-cli-syntax.js b/test/parallel/test-cli-syntax.js
index 059e3ef..20fdfdc 100644
--- a/test/parallel/test-cli-syntax.js
+++ b/test/parallel/test-cli-syntax.js
@@ -6,7 +6,7 @@ const path = require('path');
 
 const common = require('../common');
 
-var node = process.argv[0];
+var node = process.execPath;
 
 // test both sets of arguments that check syntax
 var syntaxArgs = [
@@ -49,9 +49,13 @@ var syntaxArgs = [
     var _args = args.concat(file);
     var c = spawnSync(node, _args, {encoding: 'utf8'});
 
-    // no output should be produced
+    // no stdout should be produced
     assert.equal(c.stdout, '', 'stdout produced');
-    assert(c.stderr, 'stderr not produced');
+
+    // stderr should have a syntax error message
+    var match = c.stderr.match(/^SyntaxError: Unexpected identifier$/m);
+    assert(match, 'stderr incorrect');
+
     assert.equal(c.status, 1, 'code == ' + c.status);
   });
 });
@@ -68,9 +72,13 @@ var syntaxArgs = [
     var _args = args.concat(file);
     var c = spawnSync(node, _args, {encoding: 'utf8'});
 
-    // no output should be produced
+    // no stdout should be produced
     assert.equal(c.stdout, '', 'stdout produced');
-    assert(c.stderr, 'stderr not produced');
+
+    // stderr should have a module not found error message
+    var match = c.stderr.match(/^Error: Cannot find module/m);
+    assert(match, 'stderr incorrect');
+
     assert.equal(c.status, 1, 'code == ' + c.status);
   });
 });
```

Fishrock123 commented:
I'd say to please use `const` when you can if you add or update a variable. :)

rvagg commented:
grf, bikeshed, can we defer the `var` vs `const` thing please @Fishrock123 & @thefourtheye? This 1/2 way state we're in is too messy and I haven't seen anything other than a vague idea amongst some collaborators that `const` is a must now.

rvagg commented:
landed @ 2e6ece4 with commit summary modification

thanks @bahamas10 and congrats on first commit, and it's a significant one too!

bahamas10 commented:
@rvagg + all. nice! thanks everyone.

jasnell commented:
Nice to see this landed :+1:

jays1204 commented:
I understand that we are done talking about this topic. However, I am writing to ask you one question. 
If test.js file has below contents then 'node test.js' works.
- test.js
```javascript
if (true) {
  return;
}
```
But 'node -c test.js' occurs error "SyntaxError: Illegal return statement". 

'return' statement works in function. But node module system wraps files in a self-executing function. so "node test.js" works but "node -c test.js" not works.
Is this okay?  
If I bother you then I'm sorry.

bnoordhuis commented:
@jays1204 I'd consider that a bug.  Can you open an issue?

jays1204 commented:
@bnoordhuis already someone opened it :)
